### ISSUE TITLE
OpenUSB backend hangs on control requests

### ISSUE THREAD

**Issue by walac** [pyusb/pyusb#60]
[OpenUSB](http://sourceforge.net/projects/openusb/) backend hangs on some control requets. [This script](https://gist.github.com/walac/8685660) reproduces the problem.

**Comment by walac** [comment]
openusb hasn't been update for 2 years, it makes no sense holding a release because of it. Unblocked.

**Comment by jonasmalacofilho** [comment]
Still reproduces with OpenUSB 1.1.11 and 1.1.14 on Linux 5.14.

_Haven't tried 1.1.16 because its sources are missing (both from VCS and as a tarball) and a random binary release after 2 years of complete inactivity seems a bit suspicious._

I did a quick strace and this seems to be a bug in OpenUSB:

```
[pid 86536] ioctl(8, USBDEVFS_SUBMITURB <unfinished ...>
[pid 86540] set_robust_list(0x7fe8e4e0b9e0, 24 <unfinished ...>
[pid 86536] <... ioctl resumed>, 0x55d71df195a0) = 0
[pid 86536] write(10, "\0", 1)          = 1
[pid 86540] <... set_robust_list resumed>) = 0
[pid 86536] futex(0x55d71dfcaab0, FUTEX_WAIT_PRIVATE, 0, NULL <unfinished ...>
[pid 86540] select(10, [6 9], [8], NULL, {tv_sec=0, tv_usec=999920}) = 2 (in [6 9], left {tv_sec=0, tv_usec=999918})
[pid 86540] read(9, "\0", 1)            = 1
[pid 86540] read(6, "\1", 1)            = 1
[pid 86540] select(10, [6 9], [8], NULL, {tv_sec=0, tv_usec=999847}) = 1 (out [8], left {tv_sec=0, tv_usec=999755})
[pid 86540] ioctl(8, USBDEVFS_REAPURBNDELAY, 0x7fe8e4e0ad00) = 0
[pid 86540] mmap(0x7fe8dc000000, 67108864, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_NORESERVE, -1, 0) = 0x7fe8dc000000
[pid 86540] mprotect(0x7fe8dc000000, 135168, PROT_READ|PROT_WRITE) = 0
[pid 86540] ioctl(8, USBDEVFS_REAPURBNDELAY, 0x7fe8e4e0ad00) = -1 EAGAIN (Resource temporarily unavailable)
[pid 86540] select(10, [6 9], [8], NULL, {tv_sec=0, tv_usec=999675}) = 0 (Timeout)
[pid 86540] ioctl(8, USBDEVFS_DISCARDURB, 0x55d71df195a0) = -1 EINVAL (Invalid argument)
[pid 86540] select(10, [6 9], [8], NULL, {tv_sec=0, tv_usec=0}) = 0 (Timeout)
[pid 86540] ioctl(8, USBDEVFS_DISCARDURB, 0x55d71df195a0) = -1 EINVAL (Invalid argument)
[pid 86540] select(10, [6 9], [8], NULL, {tv_sec=0, tv_usec=0}) = 0 (Timeout)
[pid 86540] ioctl(8, USBDEVFS_DISCARDURB, 0x55d71df195a0) = -1 EINVAL (Invalid argument)
[pid 86540] select(10, [6 9], [8], NULL, {tv_sec=0, tv_usec=0}) = 0 (Timeout)
[pid 86540] ioctl(8, USBDEVFS_DISCARDURB, 0x55d71df195a0) = -1 EINVAL (Invalid argument)
[pid 86540] select(10, [6 9], [8], NULL, {tv_sec=0, tv_usec=0}) = 0 (Timeout)
[pid 86540] ioctl(8, USBDEVFS_DISCARDURB, 0x55d71df195a0) = -1 EINVAL (Invalid argument)
[pid 86540] select(10, [6 9], [8], NULL, {tv_sec=0, tv_usec=0}) = 0 (Timeout)
[pid 86540] ioctl(8, USBDEVFS_DISCARDURB, 0x55d71df195a0) = -1 EINVAL (Invalid argument)
[pid 86540] select(10, [6 9], [8], NULL, {tv_sec=0, tv_usec=0}) = 0 (Timeout)
[pid 86540] ioctl(8, USBDEVFS_DISCARDURB, 0x55d71df195a0) = -1 EINVAL (Invalid argument)
[pid 86540] select(10, [6 9], [8], NULL, {tv_sec=0, tv_usec=0}) = 0 (Timeout)
[pid 86540] ioctl(8, USBDEVFS_DISCARDURB, 0x55d71df195a0) = -1 EINVAL (Invalid argument)
[pid 86540] select(10, [6 9], [8], NULL, {tv_sec=0, tv_usec=0}) = 0 (Timeout)
[pid 86540] ioctl(8, USBDEVFS_DISCARDURB, 0x55d71df195a0) = -1 EINVAL (Invalid argument)
[pid 86540] select(10, [6 9], [8], NULL, {tv_sec=0, tv_usec=0}) = 0 (Timeout)
[pid 86540] ioctl(8, USBDEVFS_DISCARDURB, 0x55d71df195a0) = -1 EINVAL (Invalid argument)
[pid 86540] select(10, [6 9], [8], NULL, {tv_sec=0, tv_usec=0}) = 0 (Timeout)
[pid 86540] ioctl(8, USBDEVFS_DISCARDURB, 0x55d71df195a0) = -1 EINVAL (Invalid argument)
[pid 86540] select(10, [6 9], [8], NULL, {tv_sec=0, tv_usec=0}) = 0 (Timeout)
```

I'm inclined to close this as "wontfix", as using the OpenUSB backend seems pretty pointless nowadays.  It was supposed to be an MT-safe libusb-0, but libusb-1.0 already solves that.

Also, it clearly has no adoption on Linux, and on Mac it seems to use libusb; maybe it's still relevant for Windows?

---

Related: ArchLinux [PKGBUILD for libopenusb](https://github.com/jonasmalacofilho/archlinux-libopenusb) (for now not pushed to the AUR).

**Comment by mcuee** [comment]
I will support to drop openusb backend. It is a dead project.

**Comment by jonasmalacofilho** [comment]
Debian and Ubuntu popcon stats reinforce that OpenUSB is not widely used.  I haven't checked other distros, but Arch didn't even have a package for it.

Additionally, there never was a pre-compiled package for Windows (so we safely assume it's not terribly popular there either), and there's no (current) Homebrew formula.

Let's consider the OpenUSB deprecated as of right now, and scheduled for removal in a future major PyUSB release.
